src/execution/sql/runtime_types.cpp

```diff
@@ -655,4 +658,711 @@ Timestamp Timestamp::FromYMDHMSMU(int32_t year, int32_t month, int32_t day, int3
+}
+
+template <typename T>
+void Decimal<T>::MultiplyAndSet(const Decimal<T> &input, uint32_t precision) {
+  constexpr const uint128_t bottom_mask = (uint128_t{1} << 64) - 1;
+  constexpr const uint128_t top_mask = ~bottom_mask;
+
+  // First input
+  uint128_t a = value_;
+  // Second input
+  uint128_t b = input.GetValue();
+  // Split into half words
+  uint128_t half_words_a[2];
+  uint128_t half_words_b[2];
+
+  half_words_a[0] = a & bottom_mask;
+  half_words_a[1] = (a & top_mask) >> 64;
+
+  half_words_b[0] = b & bottom_mask;
+  half_words_b[1] = (b & top_mask) >> 64;
+
+  // Calculate 256 bit result
+  uint128_t half_words_result[4];
+  CalculateMultiWordProduct128(half_words_a, half_words_b, half_words_result, 2, 2);
+
+  if (half_words_result[2] == 0 && half_words_result[3] == 0) {
+    // TODO(Rohan): Optimize by sending in an array of half words
+    value_ = half_words_result[0] | (half_words_result[1] << 64);
+    UnsignedDivideConstant128BitPowerOfTen(precision);
+    return;
+  }
+
+  // Magic number halfwords
+  uint128_t magic[4];
+  magic[0] = MAGIC_ARRAY[precision][3];
+  magic[1] = MAGIC_ARRAY[precision][2];
+  magic[2] = MAGIC_ARRAY[precision][1];
+  magic[3] = MAGIC_ARRAY[precision][0];
+
+  uint32_t magic_p = MAGIC_P_AND_ALGO_ARRAY[precision][0] - 256;
+
+  if (MAGIC_P_AND_ALGO_ARRAY[precision][1] == 0) {
+    // Overflow Algorithm 1 - Magic number is < 2^256
+
+    // Magic Result
+    uint128_t half_words_magic_result[8];
+    // TODO(Rohan): Make optimization to calculate only upper half of the word
+    CalculateMultiWordProduct128(half_words_result, magic, half_words_magic_result, 4, 4);
+    // Get the higher order result
+    uint128_t result_lower = half_words_magic_result[4] | (half_words_magic_result[5] << 64);
+    uint128_t result_upper = half_words_magic_result[6] | (half_words_magic_result[7] << 64);
+
+    uint128_t overflow_checker = result_upper >> magic_p;
+    if (overflow_checker > 0) {
+      // Result will overflow from 128 bits
+      throw EXECUTION_EXCEPTION(fmt::format("Result overflow > 128 bits"), common::ErrorCode::ERRCODE_DATA_EXCEPTION);
+    }
+
+    result_lower = result_lower >> magic_p;
+    result_upper = result_upper << (128 - magic_p);
+    value_ = result_lower | result_upper;
+    return;
+  }
+  // Overflow Algorithm 2 - Magic number is > 2^256
+
+  // Magic Result
+  uint128_t half_words_magic_result[8];
+  // TODO(Rohan): Make optimization to calculate only upper half of the word
+  CalculateMultiWordProduct128(half_words_result, magic, half_words_magic_result, 4, 4);
+  // Get the higher order result
+  uint128_t result_lower = half_words_result[0] | (half_words_result[1] << 64);
+  uint128_t result_upper = half_words_result[2] | (half_words_result[3] << 64);
+
+  uint128_t add_lower = half_words_magic_result[4] | (half_words_magic_result[5] << 64);
+  uint128_t add_upper = half_words_magic_result[6] | (half_words_magic_result[7] << 64);
+
+  /*Perform addition*/
```

Comment:
Nit: the inconsistent commenting is getting to me a bit, but we can clean that up